src/mod/internal/login_mod.cpp

```diff
@@ -27,6 +27,31 @@
 #include "core/RDP/slowpath.hpp"
 #include "RAIL/client_execute.hpp"
 
+#include <string>
+#include <tuple>
+
+namespace
+{
+    inline std::tuple<std::string, std::string> rpartition(std::string text,
+                                                           const char* seps) {
+        size_t sep_pos = text.find_last_of(seps);
+        if (sep_pos != std::string::npos) {
+            return {text.substr(0, sep_pos), text.substr(sep_pos + 1)};
+        }
+        return {std::string(), text};
+    }
+
+    inline std::string concat_target_login(std::string login, std::string target) {
+        if (target.empty()) {
+            return login;
+        }
+        size_t sep_pos = target.find_last_of("+:");
+        if (sep_pos != std::string::npos) {
+            return target + target[sep_pos] + login;
```

Comment:
`str_concat(...)` (`utils/sugar/algostring.hpp`) is preferable to multi-concatenation.